cpp/src/Ice/Protocol.cpp

```diff
@@ -3,9 +3,97 @@
 #include "Protocol.h"
 #include "Ice/LocalExceptions.h"
 
+#include <sstream>
+
 using namespace std;
 using namespace Ice;
 
+namespace
+{
+    template<typename T> std::string versionToString(const T& v)
+    {
+        std::ostringstream os;
+        os << v;
+        return os.str();
+    }
+
+    template<typename T> T stringToVersion(std::string_view str)
+    {
+        T v;
+        IceInternal::stringToMajorMinor(str, v.major, v.minor);
+        return v;
+    }
+}
+
+string
+IceInternal::protocolVersionToString(const ProtocolVersion& v)
+{
+    return versionToString<ProtocolVersion>(v);
+}
+
+ProtocolVersion
+IceInternal::stringToProtocolVersion(string_view v)
+{
+    return stringToVersion<ProtocolVersion>(v);
+}
+
+string
+IceInternal::encodingVersionToString(const EncodingVersion& v)
+{
+    return versionToString<EncodingVersion>(v);
+}
+
+EncodingVersion
+IceInternal::stringToEncodingVersion(string_view v)
+{
+    return stringToVersion<EncodingVersion>(v);
+}
+
+ostream&
+Ice::operator<<(ostream& out, const ProtocolVersion& version)
+{
+    return out << static_cast<int>(version.major) << "." << static_cast<int>(version.minor);
+}
+
+ostream&
+Ice::operator<<(ostream& out, const EncodingVersion& version)
+{
+    return out << static_cast<int>(version.major) << "." << static_cast<int>(version.minor);
+}
+
+void
+IceInternal::stringToMajorMinor(string_view s, uint8_t& major, uint8_t& minor)
+{
+    string str{s};
```

Comment:
We can consider validating versions only contains decimal digits and one dot. I think the integer conversion here will allow a sign, and non decimal base numbers.

But it might be overkill as this is now internal and we should not be passing such version strings.